samcli/commands/local/lib/intrinsic_resolver/intrinsic_property_resolver.py

```diff
@@ -0,0 +1,759 @@
+        the easier resolution will be.
+
+        In the future, for items like Fn::ImportValue multiple templates can be provided
+        into the function.
+        Parameters
+        -----------
+        symbol_resolver: IntrinsicsSymbolTable
+            This is a required property that is the crux of the intrinsics resolver. It acts as a mapping between
+            resource_type and object to the Ref, Arn, etc.
+        template: dict
+            A dictionary containing all the resources and the attributes of the CloudFormation template.
+            Customers can pass in the template or the individual properties like resources, mapping, parameters,
+            conditions.
+        resources: dict
+            Dictionary containing the CloudFormation resources
+        mappings: dict
+            Dictionary containing the CloudFormation mappings. This is used in the Fn::FindInMap section
+        parameters: dict
+            Dictionary containing the CloudFormation parameters. This is used in the Ref section
+        conditions: dict
+            Dictionary containing the CloudFormation conditions. This is used in the Boolean Intrinsics section
+        """
+        self.template = template or {}
+        self.resources = resources or self.template.get("Resources", {})
+        self.mapping = mappings or self.template.get("Mappings", {})
+        self.parameters = parameters or self.template.get("Parameters", {})
+        self.conditions = conditions or self.template.get("Conditions", {})
+
+        self.symbol_resolver = symbol_resolver
+        self.symbol_resolver.resources = self.resources
+        self.symbol_resolver.parameters = self.parameters
+
+        self.intrinsic_key_function_map = {
+            IntrinsicResolver.FN_JOIN: self.handle_fn_join,
+            IntrinsicResolver.FN_SPLIT: self.handle_fn_split,
+            IntrinsicResolver.FN_SUB: self.handle_fn_sub,
+            IntrinsicResolver.FN_SELECT: self.handle_fn_select,
+            IntrinsicResolver.FN_BASE64: self.handle_fn_base64,
+            IntrinsicResolver.FN_FIND_IN_MAP: self.handle_find_in_map,
+            IntrinsicResolver.FN_TRANSFORM: self.handle_fn_transform,
+            IntrinsicResolver.FN_GET_AZS: self.handle_fn_get_azs,
+            IntrinsicResolver.REF: self.handle_fn_ref,
+            IntrinsicResolver.FN_GET_ATT: self.handle_fn_getatt,
+            IntrinsicResolver.FN_IMPORT_VALUE: self.handle_fn_import_value
+        }
+
+        self.conditional_key_function_map = {
+            IntrinsicResolver.FN_AND: self.handle_fn_and,
+            IntrinsicResolver.FN_OR: self.handle_fn_or,
+            IntrinsicResolver.FN_IF: self.handle_fn_if,
+            IntrinsicResolver.FN_EQUALS: self.handle_fn_equals,
+            IntrinsicResolver.FN_NOT: self.handle_fn_not
+        }
+
+    def intrinsic_property_resolver(self, intrinsic, parent_function="template"):
```

Comment:
Having a hard time understanding what `parent_function="template"` is.

Reply:
This simplifies figuring out where an error has occured since this function is called generically to provide a better error message. By default, when parsing the template the error will message will assume the location is the "template". However, inside nested Intrinsics having something like self.intrinsic_property_resolver(intrinsic_value, parent_function=IntrinsicResolver.FN_JOIN) simplifies the handling by telling users their errors are in Fn::Join. 

This happens in two cases. The object property resolved is missing or it is a nested dictionary within the object that needs to be resolved.